test/LocalVocabTest.cpp

```diff
@@ -20,23 +34,146 @@ TEST(LocalVocab, construction) {
+  Values values1(testQec, {{Variable{"?x"}, Variable{"?y"}},
+                           {{TripleComponent{"x"}, TripleComponent{"y1"}},
+                            {TripleComponent{"x"}, TripleComponent{"y2"}}}});
+  checkLocalVocab(&values1, std::vector<std::string>{"x", "y1", "y2"});
+
+  // VALUES operation that uses an existing literal (from the test index). Note
+  // that the quotes are part of the name of the literals.
+  Values values2(testQec, {{Variable{"?x"}}, {{TripleComponent{"\"alpha\""}}}});
+  checkLocalVocab(&values2, std::vector<std::string>{});
+
+  // JOIN operation with exactly one non-empty local vocab.
+  Join join1(testQec, qet(values1), qet(values2), 1, 1);
+  checkLocalVocab(&join1, std::vector<std::string>{"x", "y1", "y2"});
+
+  // JOIN operation with two non-empty local vocab.
+  Join join2(testQec, qet(values1), qet(values1), 1, 1);
+  checkThrow(&join2);
+
+  // ORDER BY operation.
+  // Note: the third argument ar the indices of the columns to be sorted, and
+  // the sort order. It's not important for this test.
+  // TODO: Is the note correct? Unfortunately, `OrderBy.h` is not documented.
+  OrderBy orderBy(testQec, qet(values1), {{0, true}, {1, true}});
+  checkLocalVocab(&orderBy, std::vector<std::string>{"x", "y1", "y2"});
+
+  // DISTINCT operation.
+  // Note: the third argument are those indices of the input columns that are
+  // considered for the output.
+  // TODO: Is the note correct? Unfortunately, `Distinct.h` is not documented.
+  Distinct distinct(testQec, qet(values1), {0, 1});
+  checkLocalVocab(&distinct, std::vector<std::string>{"x", "y1", "y2"});
+
+  // GROUP BY operation.
+  auto groupConcatExpression =
+      [](std::string variableName,
+         std::string separator) -> sparqlExpression::SparqlExpressionPimpl {
+    // NOTE: The last argument is a string representation of the expression,
+    // which does not matter for the purposes of this test.
+    return {std::make_unique<sparqlExpression::GroupConcatExpression>(
+                false,
+                std::make_unique<sparqlExpression::VariableExpression>(
+                    Variable{variableName}),
+                separator),
+            "GROUP_CONCAT"};
+  };
+  GroupBy groupBy(
+      testQec, {Variable{"?x"}},
+      {Alias{groupConcatExpression("?y", "|"), Variable{"?concat"}}},
+      qet(values1));
+  checkLocalVocab(&groupBy, std::vector<std::string>{"x", "y1", "y2", "y1|y2"});
+
+  // DISTINCT again, but after something has been added to the local vocabulary.
+  //
+  // TODO: Without a clear cache (see `checkLocalVocab` above), the "y1|y2" from
+  // the previous GROUP BY operation remains in the local vocabulary. This is
+  // surprising behavior for me because I thought the local vocab is tied to the
+  // result table (which is created from scratch in each call to
+  // `checkLocalVocab`) and not to the operation (`values1` in this case).
+  // `
```

Comment:
I know that it is unexpected, but it is a consequence of how we implemented the handling of the local vocab inside the `GroupBy`. The group by adds its local vocab to the local vocab of the subtree. For the sake of this unit test, this is unexpected. But it is also possibly a race condition, because values inside the cache might change after they have been added and made accessible to other queries. We should definitely discuss how we should handle the interaction between local vocabs and the cache.